### PR TITLE
label based prometheus traffic stats

### PR DESCRIPTION
**Pull request by eberhab**

### CODE REVIEW

File: hopglass-server.js [hunk 1]
```diff
@@ -563,11 +592,20 @@ function getMetrics(stream) {
       stream.write('meshnodes_total ' + counter_meshnodes_total + '\n')
       stream.write('meshnodes_online_total ' + counter_meshnodes_online_total + '\n')
       stream.write('total_clients ' + counter_clients + '\n')
-      stream.write('total_traffic_rx ' + counter_traffic_rx + '\n')
-      stream.write('total_traffic_mgmt_rx ' + counter_traffic_mgmt_rx + '\n')
-      stream.write('total_traffic_tx ' + counter_traffic_tx + '\n')
-      stream.write('total_traffic_mgmt_tx ' + counter_traffic_mgmt_tx + '\n')
-      stream.write('total_traffic_forward ' + counter_traffic_forward + '\n')
+
+      var labels = [];
```
Comment (plumpudding) [inline hopglass-server.js]: I wonder how this works. You can't have a String as index for an array.
You should be using an Object ({}) here. 
If you don't mind I will redo this.

**Comment by plumpudding** [comment]
As this is the master branch I'll just take into account that it works. I will do a port to the modules branch.

**Comment by eberhab** [comment]
interesting, i must have confused this with python :) I just looked into this and the reason this works seems to be that js automatically converts the associative array into an object:

http://www.w3schools.com/js/js_arrays.asp -> Associative Arrays -> Warning

thank you for porting this to the modules branch

File: hopglass-server.js [hunk 2]
```diff
@@ -480,9 +480,24 @@ function getFfmapJson(stream) {
 
 function getMetrics(stream) {
   getData()
-  var save = (n, id, stream, what, where) => {
-    if (_.has(n, what))
-      stream.write((where ? where : what.replace(/\./g, '_')) + id + ' ' +  _.get(n, what) + '\n')
+  var save = (n, stream, labels, what, value) => {
```
Comment (plumpudding) [inline hopglass-server.js]: What's the difference between "what" and "value"?

File: hopglass-server.js [hunk 3]
```diff
@@ -480,9 +480,24 @@ function getFfmapJson(stream) {
...
+    var id = '{'
+    var first = true
+    for (var e in labels) {
+      if (first)
+        first = false
+      else
+        id += ','
+      id += e + '="' + labels[e] + '"'
+    }
+    id += "}"
```
Comment (plumpudding) [inline hopglass-server.js]: easier (at least for me):

      var newLabels = []
      Object.keys(labels).map(function(key) {
        newLabels.push(key + '="' + labels[key] + '"')
      })
      labels = '{' + newLabels.join(',') + '}'

File: hopglass-server.js [hunk 4]
```diff
@@ -512,18 +527,32 @@ function getMetrics(stream) {
...
-        save(n, id, stream, 'statistics.traffic.mgmt_rx.bytes')
-        save(n, id, stream, 'statistics.traffic.tx.bytes')
-        save(n, id, stream, 'statistics.traffic.mgmt_tx.bytes')
-        save(n, id, stream, 'statistics.traffic.forward.bytes')
-        save(n, id, stream, 'statistics.loadavg')
-        if (_.has(n, 'statistics.memory.free') && _.has(n, 'statistics.memory.total'))
-          stream.write('statistics_memory_usage' + id + ' ' + (n.statistics.memory.total - n.statistics.memory.free)/n.statistics.memory.total + '\n')
+      if (_.has(n, 'nodeinfo.hostname') && _.has(n, 'statistics.gateway')) {
+
+        var labels = [];
```
Comment (plumpudding) [inline hopglass-server.js]: we don't use semicolons!

File: hopglass-server.js (hunk 2 above)
Comment (eberhab) [inline hopglass-server.js]: Well, "what" is actually exactly the same as it was originally, which is what I'd call the metrics key from the statistics json. It is also used to form the prometheus metric name. 

And "value" is its value which is to be the value of the prometheus metric.

Also I tried to conserve the original functionality of the "save" function in a way that "what" can be used as both:
- as a key to the statistics dict in order to extract a value
- as a prometheus metrics name
the former being overruled by specifying a an optional value argument to the save function.

File: hopglass-server.js (hunk 3 above)
Comment (eberhab) [inline hopglass-server.js]: I guess this is the way it should be. The reason it worked before was probably just that js compensated for my language ignorance :)

File: hopglass-server.js (hunk 4 above)
Comment (eberhab) [inline hopglass-server.js]: ack

**Pull request merged**